lightning/src/ln/peers/conduit.rs

```diff
@@ -0,0 +1,173 @@
+use ln::peers::{chacha, hkdf};
+
+/// Returned after a successful handshake to encrypt and decrypt communication with peer nodes
+pub struct Conduit {
+	pub(crate) sending_key: [u8; 32],
+	pub(crate) receiving_key: [u8; 32],
+
+	pub(crate) sending_chaining_key: [u8; 32],
+	pub(crate) receiving_chaining_key: [u8; 32],
+
+	pub(crate) receiving_nonce: u32,
+	pub(crate) sending_nonce: u32,
+
+	pub(super) read_buffer: Option<Vec<u8>>,
+}
+
+impl Conduit {
+	pub fn encrypt(&mut self, buffer: &[u8]) -> Vec<u8> {
+		let length = buffer.len() as u16;
+		let length_bytes = length.to_be_bytes();
+
+		let encrypted_length = chacha::encrypt(&self.sending_key, self.sending_nonce as u64, &[0; 0], &length_bytes);
+		self.increment_sending_nonce();
+
+		let encrypted_message = chacha::encrypt(&self.sending_key, self.sending_nonce as u64, &[0; 0], buffer);
+		self.increment_sending_nonce();
+
+		let mut ciphertext = encrypted_length;
+		ciphertext.extend_from_slice(&encrypted_message);
+		ciphertext
+	}
+
+	pub(super) fn read(&mut self, data: &[u8]) {
+		let mut read_buffer = if let Some(buffer) = self.read_buffer.take() {
```

Comment:
This would probably be more readable with https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert

Reply:
Ooooh, TIL. There are a bunch of places where I will replace that.